### DISCUSSION TITLE
Is nicegui app stateful?

### DISCUSSION THREAD

**Discussion by AsiaCao** [zauberzeug/nicegui#1223] [Q&A] [answered]
### Question

Hi team,

I just have two questions:

1. Is a `nicegui` app stateful by default or is there any feature could make the app stateful? 
by stateful, i refer to the server has to store any user-specific/session-specific information such that subsequent requests from the same user has to go to the same server.

    **More context**:
    We have a web application that is built with Streamlit. The app is served by a fleet of servers (2+) behind a reverse proxy. We find that a streamlit application can be stateful with the session state feature. Due to session state, we have to implement a sticky session mechanism to ensure the request goes to the right server. This is not ideal for scalability. So I'm wondering if niceui have similiar behaviors or features?

2. do you know any documentations about how nicegui works internally? like how does it detect UI state changes, how does frontend communicates with the backend, what kind protocols are used, etc. These kind of documentation would be helpful for tech savvy community members to learn more about nicegui and be more self-service. Thanks!

**Comment by rodja** [comment, marked as answer]
Yes, NiceGUI is stateful by design. The initially loaded html page connects back to the backend via websocket to deliver user events and receive content updates. This websocket connection must be routed to the same server which delivered the page (including all possible reconnections). As you said, this is not ideal for scalability. But there is (almost) no way around it because client-state is only available on the specific backend server. The only trick you could pull is to save all state in a session cookie on the client (`app.storage.browser`) as we did to toggle the dark mode for the https://nicegui.io website: 

https://github.com/zauberzeug/nicegui/blob/6f81f3241777a9d87edc24ea12fcb2eb8eeca2a5/main.py#L82-L88

**Reply by AsiaCao** [reply]
got it, thank you for your quick response and the detailed explanation. very helpful, thanks @rodja!
